lock-api/src/main/java/com/palantir/lock/client/SharedLockToken.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import com.google.common.base.Preconditions;
+import com.palantir.lock.v2.LockToken;
+
+class SharedLockToken implements LockToken {
+    private final UUID requestId;
+    private final ReferenceCounter referenceCounter;
+
+    private volatile boolean unlocked;
+
+    private SharedLockToken(ReferenceCounter referenceCounter) {
+        this.referenceCounter = referenceCounter;
+        this.requestId = UUID.randomUUID();
+        this.unlocked = false;
+    }
+
+    public static List<LockToken> share(LockToken token, int referenceCount) {
```

Comment:
being a bit pedantic here, but what happens if you share a `SharedLockToken`?

i.e. if I have raw lock token, share it 5 times, and then for one of the shared tokens I share that 3 times, but then I go and unlock the first 5 shared lock tokens. What "guarantees" do I have about holding a `SharedLockToken`?

Aside from guarantees or whatever, it feels like it should special case `SharedLockToken` and pipe through the underlying lock token + somehow use the same reference counter?

If this won't come up in actual use case, then probably not worth it, but in terms of what guarantees you get holding a `SharedLockToken`, something feels amiss. Happy to be wrong though.

Reply:
we won't be sharing a shared lock token, but I think we can explicitly check that. That is, throw if someone tries to create shared lock tokens from another shared lock token.